pepr/privacy/mia.py

```diff
@@ -0,0 +1,1182 @@
+                fp = np.count_nonzero((pred == True) & (true_labels == False))
+                precision = tp / (tp + fp) if (tp + fp) else 1
+                recall = tp / (fn + tp) if (fn + tp) else 0
+                test_accuracy = np.count_nonzero(pred == true_labels) / len(pred)
+
+                # Store evaluation results to matrices
+                tn_list[am][tm] = tn
+                tp_list[am][tm] = tp
+                fn_list[am][tm] = fn
+                fp_list[am][tm] = fp
+                precision_list[am][tm] = precision
+                recall_list[am][tm] = recall
+                test_accuracy_list[am][tm] = test_accuracy
+
+        # Average over all attack models per target model
+        precision = np.average(precision_list, axis=0)
+        recall = np.average(recall_list, axis=0)
+        test_accuracy = np.average(test_accuracy_list, axis=0)
+
+        # Average over all attack models and target models
+        precision_all = np.average(precision)
+        recall_all = np.average(recall)
+        test_accuracy_all = np.average(test_accuracy)
+
+        attack_model_results = {
+            "tn_list": tn_list,
+            "tp_list": tp_list,
+            "fn_list": fn_list,
+            "fp_list": fp_list,
+            "precision_list": precision_list,
+            "recall_list": recall_list,
+            "test_accuracy_list": test_accuracy_list,
+            "test_accuracy": test_accuracy,
+            "precision": precision,
+            "recall": recall,
+            "overall_test_accuracy": test_accuracy_all,
+            "overall_precision": precision_all,
+            "overall_recall": recall_all,
+        }
+
+        return attack_model_results
+
+    @staticmethod
+    def _shadow_model_evaluation(
+        shadow_models,
+        shadow_data_indices,
+        shadow_data,
+        shadow_labels,
+    ):
+        """
+        Evaluate shadow models.
+
+        Parameters
+        ----------
+        shadow_models : iterable
+            List of trained shadow models to evaluate.
+        shadow_data_indices : numpy.ndarray
+            Array of training and evaluation indices for the shadow models
+        shadow_data : numpy.ndarray
+            Training and evaluation data for the shadow models.
+        shadow_labels : numpy.ndarray
+            Training and evaluation labels for the shadow models.
+
+        Returns
+        -------
+        dict
+            Dictionary storing the shadow model results.
+
+            * shadow_train_accuracy_list (list): Accuracy on train data per shadow
+              model.
+            * shadow_test_accuracy_list (list): Accuracy on evaluation data per shadow
+              model.
+            * shadow_train_accuracy (float): Accuracy on train records averaged over all
+              shadow models.
+            * shadow_test_accuracy (float): Accuracy on evaluation records averaged over
+              all shadow models.
```

Comment:
Also here better ...`evaluation`... than ...`test`...?

Reply:
Changed in 03c76ef